### ISSUE TITLE
Memory saving only in checkpoint size, not during training

### ISSUE THREAD

**Issue by aditya2331** [zyushun/Adam-mini#8]
Hi, I tried running the normal train_gpt2.py code with Adam_mini. I had to remove the ValueError and Hessian spectrum import to make it work properly. I noticed that there was a good reduction in the checkpoint size but no memory reduction during the training, so I couldn't fit a higher batch size for more throughput. The memory usage came out to be the same as regular Adam while training.

Training Details:

DDP with 4 GPUs, batch size 12, 336 gradient accumulation steps

**Comment by chcoliang** [comment]
Since the checkpoint size is reduced, the optimizer state should be reduced. According to the model size, there should be about 5G memory reduction when switching from AdamW to Adam-mini. If no memory reduction occurs, you can check whether you offload the optimizer to the CPU.

To fit larger batch size,  as Adam-mini can only save about 5G, we recommend cooperating with other memory-saving techniques, such as gradient checking and CPU offloading.

**Comment by aditya2331** [comment]
The first 4 GPUs are using adam mini, the latter are using adamw. Interestingly, switching to adamw on your codebase runs into an OOM error, but using the original nanoGPT runs fine.

![Image 08-07-24 at 11 44 AM](https://github.com/zyushun/Adam-mini/assets/92774770/eeb72f86-f150-4488-89b6-25c549101c48)  

As can be seen, one GPU is using more memory than the others in adam mini, not sure why. This is the original GPT2 XL model without weight tying. Batch size and gradient accumulation are same as those above (12, 336).

There is no CPU offloading.

**Comment by zyushun** [comment]
Hi @aditya2331 ! Thanks for the update!

Regarding your figure: it seems that there is only ~300m  mem reduction on each GPU. This is unexpected because normally it should be ~5GB reduction in your setting. Could you double check your ckpt size to see if there is actually ~5GB reduction?  

Regarding your comment "adamw on your codebase runs into an OOM error, but using the original nanoGPT runs fine." : this also seems unusual since our code is essentially nanoGPT.  Perhaps a simple way debug is to: import and run Adam-mini on the original nanoGPT code to see if things get better.

Please feel free to  further update here!

**Comment by 980202006** [comment]
Same issue in huggingface trainer.

**Comment by 980202006** [comment]
Same issue in huggingface trainer.

**Comment by zyushun** [comment]
Hi @980202006 ! This issue seems unusual. We do observe substantial memory reduction on Huggingface trainer. We share our results as follows. 

Setting: We use Huggingface trainer under [Llamafactory](https://github.com/hiyouga/LLaMA-Factory) codebase.  We conduct  SFT on Llama2-7b.  We use gradient checkpointing, batch size  = 4 and DeepSpeed zero3. 

The memory usage is shown as follows:
```
Adam-mini： 34435.0MB  35533.0MB  36701.0MB  35207.0MB

Adamw： 49027.0MB 48985.0MB 48991.0MB 48439.0MB
```

Could you share more training details on your side? It would help us debug. Thanks a lot!

**Comment by 980202006** [comment]
@zyushun Thank you! I will try again.

**Comment by zyushun** [comment]
Hi @aditya2331 , regarding " switching to adamw on your codebase runs into an OOM error, but using the original nanoGPT runs fine."

I found it might due to different attention implementation between our [model.py](https://github.com/zyushun/Adam-mini/blob/main/gpt2/model.py) (which is actually an old version from nanoGPT) and the latest [model.py in nanoGPT](https://github.com/karpathy/nanoGPT/blob/master/model.py).  The difference lies in "CausalSelfAttention": the old version uses separated Q, K, V while the latest one combines them together. Mathematically, these two implementations should be equivalent, but computationally, they might lead to different memory consumption. 

We recommend using  our [model.py](https://github.com/zyushun/Adam-mini/blob/main/gpt2/model.py) to avoid any potential unexpected error.
